File: modules/web/js/ballerina/utils/common-utils.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
define(['lodash', 'log'], function (_, log) {

    var CommonUtils = function () {
    };

    /**
     * Generates a unique id for an attribute by checking against a list of parents.
     *
     * @param {Object} genArgs - Arguments for creating an object.
     * @param {ASTNode} genArgs.node - The AST node of which the attribute belongs to.
     * @param {Object[]} genArgs.attributes[] - The list of attributes to which the unique ID to be generated.
     * @param {string} genArgs.attributes[].defaultValue - The default value for the attribute.
     * @param {function} genArgs.attributes[].setter - The setter function of the attribute.
     * @param {function} genArgs.attributes[].getter - The getter function of the attribute.
     * @param {Object[]} genArgs.attributes[].parents[] - The list of parents to traverse through to generate a unique
     * ID.
     * @param {ASTNode} genArgs.attributes[].parents[].node - The AST node of the parent.
     * @param {function} genArgs.attributes[].parents[].getChildrenFunc - The getter function to get the children of a
     * parent node.
     * @param {function} genArgs.attributes[].parents[].getter - The getter function which returns the identifier of the
     * children returned from getChildrenFunc.
     */
    CommonUtils.generateUniqueIdentifier = function (genArgs) {
        _.forEach(genArgs.attributes, function (attribute) {
            if (_.isNil(attribute.getter.call(genArgs.node)) || attribute.checkEvenIfDefined) {
                // To store all the identifiers of the parents.
                var existingIdentifiers = [];

                _.forEach(attribute.parents, function (parent) {
                    log.debug("Children: " + parent.getChildrenFunc.call(parent.node));
                    // Get the children of the parent.
                    _.forEach(parent.getChildrenFunc.call(parent.node), function (child) {
                        // Skipping the current node.
                        if (!_.isEqual(genArgs.node.getID(), child.getID())) {
                            var childVal = parent.getter.call(child);
                            existingIdentifiers.push(childVal);
                        }
                    });
                });

                log.debug("Existing identifiers: " + existingIdentifiers);

                // Generating the ID.
                var counter = 0;
                var currentAttributeValue = attribute.defaultValue;
                while (true) {
                    var tempNewValue = counter > 0 ? currentAttributeValue + counter : currentAttributeValue;
                    if (!_.includes(existingIdentifiers, tempNewValue)) {
                        break;
                    }
                    counter++
                }

                if (_.isEqual(counter, 0)) {
                    attribute.setter.call(genArgs.node, currentAttributeValue);
                } else {
                    attribute.setter.call(genArgs.node, currentAttributeValue + counter);
                }
            }
        })
    };

    return CommonUtils;
});

File: misc/testerina/modules/testerina-core/src/main/java/org/ballerinalang/testerina/natives/CommonUtils.java
// Copyright (c) 2023 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
//
// WSO2 Inc. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.ballerinalang.testerina.natives;

import io.ballerina.runtime.api.types.FunctionType;
import io.ballerina.runtime.api.types.Parameter;
import io.ballerina.runtime.api.types.Type;
import io.ballerina.runtime.api.types.UnionType;
import io.ballerina.runtime.api.values.BDecimal;
import io.ballerina.runtime.api.values.BFunctionPointer;

/**
 * Common utility functions for the Testerina module.
 *
 * @since 2201.9.0
 */
public class CommonUtils {

    /**
     * Get the current time in milliseconds.
     *
     * @return The current time in milliseconds
     */
    public static BDecimal currentTimeInMillis() {
        return BDecimal.valueOf(System.currentTimeMillis());
    }

    /**
     * Check whether the parameters of a function are concurrency safe.
     *
     * @param func The function pointer
     * @return Whether the parameters are concurrency safe
     */
    public static Object isFunctionParamConcurrencySafe(BFunctionPointer func) {
        FunctionType functionType = (FunctionType) func.getType();
        Parameter[] functionParameters = functionType.getParameters();
        for (Parameter functionParameter : functionParameters) {
            Type parameterType = functionParameter.type;
            if (!isSubTypeOfReadOnly(parameterType)) {
                return false;
            }
        }
        return true;
    }

    private static boolean isSubTypeOfReadOnly(Type type) {
        if (type.isReadOnly()) {
            return true;
        }

        if (!(type instanceof UnionType unionType)) {
            return false;
        }

        for (Type memberType : unionType.getMemberTypes()) {
            if (!isSubTypeOfReadOnly(memberType)) {
                return false;
            }
        }
        return true;
    }
}
